loki/transformations/parallel/openmp_region.py

```diff
@@ -0,0 +1,271 @@
+from loki.types import DerivedType
+
+
+__all__ = [
+    'remove_openmp_regions', 'add_openmp_regions',
+    'remove_explicit_firstprivatisation',
+    'create_explicit_firstprivatisation'
+]
+
+
+def remove_openmp_regions(routine, insert_loki_parallel=False):
+    """
+    Remove any OpenMP parallel annotations (``!$omp parallel``).
+
+    Optionally, this can replace ``!$omp parallel`` with ``!$loki
+    parallel`` pragmas.
+
+    Parameters
+    ----------
+    routine : :any:`Subroutine`
+        The routine from which to strip all OpenMP annotations.
+    insert_loki_parallel : bool
+        Flag for the optional insertion of ``!$loki parallel` pragmas
+    """
+
+    class RemoveOpenMPRegionTransformer(Transformer):
+        """
+        Remove OpenMP pragmas from "parallel" regions and remove all
+        contained OpenMP pragmas and pragma regions.
+
+        Optionally replaces outer ``!$omp parallel`` region with
+        ``!$loki parallel`` region.
+        """
+
+        def visit_PragmaRegion(self, region, **kwargs):
+            """
+            Perform the fileterin and removeal of OpenMP pragma regions.
+
+            Parameters
+            ----------
+            active : tuple
+                Flag to indicate whether we're actively traversing an
+                outer OpenMP region.
+            """
+            if not region.pragma.keyword.lower() == 'omp':
+                return region
+
+            if kwargs['active'] and region.pragma.keyword.lower() == 'omp':
+                # Remove other OpenMP pragam regions when in active mode
+                region._update(pragma=None, pragma_post=None)
+                return region
+
+            if 'parallel' in region.pragma.content.lower():
+                # Replace or remove pragmas
+                pragma = None
+                pragma_post = None
+                if insert_loki_parallel:
+                    pragma = ir.Pragma(keyword='loki', content='parallel')
+                    pragma_post = ir.Pragma(keyword='loki', content='end parallel')
+
+                with dict_override(kwargs, {'active': True}):
+                    body = self.visit(region.body, **kwargs)
+
+                region._update(body=body, pragma=pragma, pragma_post=pragma_post)
+
+            return region
+
+        def visit_Pragma(self, pragma, **kwargs):
+            """ Remove other OpenMP pragmas if in active region """
+
+            if kwargs['active'] and pragma.keyword.lower() == 'omp':
+                return None
+
+            return pragma
+
+    with pragma_regions_attached(routine):
+        routine.body = RemoveOpenMPRegionTransformer().visit(routine.body, active=False)
+
+
+def add_openmp_regions(routine, global_variables=None, field_group_types=None):
+    """
+    Add the OpenMP directives for a parallel driver region with an
+    outer block loop.
+    """
+    block_dim_size = 'YDGEOMETRY%YRDIM%NGPBLKS'
+
+    global_variables = global_variables or {}
```

Comment:
As I understand it, these are the variables that compilers struggle to firstprivatise, so perhaps it should be renamed to `skip_firstprivatise_vars` or something similar. `global_variables` should be reserved for module imports. And could the tests also be extended to cover this functionality?

Reply:
Agreed that "global" is ambiguous in this context, but `skip_firstprivate_vars` is also to long and unclear (they can also not be "private"). How about `shared_variables` to stick with OpenMP semantics?